src/main/kotlin/org/rust/ide/utils/ExtractUtils.kt

```diff
@@ -0,0 +1,290 @@
+        val parameterToBounds = mutableMapOf<String, MutableSet<String>>()
+        val lifetimeToBounds = mutableMapOf<String, MutableSet<String>>()
+
+        fun normalizePredicate(text: String, name: String): String = text.removePrefix("$name:").trim()
+        fun addIfMissing(map: MutableMap<String, MutableSet<String>>, key: String) {
+            if (key !in map) {
+                map[key] = mutableSetOf()
+            }
+        }
+
+        loop@ for (predicate in wherePredList) {
+            val typeRef = predicate.typeReference
+            val lifetime = predicate.lifetime
+            when {
+                // type bound
+                typeRef != null && hasTypeParameter(typeRef, parameterMap) -> {
+                    val parameterText = (predicate.forLifetimes?.text?.plus(" ") ?: "") + typeRef.text
+                    addIfMissing(parameterToBounds, parameterText)
+                    predicate.typeParamBounds?.polyboundList?.forEach {
+                        parameterToBounds[parameterText]?.add(it.text)
+                    }
+                }
+                // lifetime bound
+                lifetime != null -> {
+                    val lifetimePredicate = createLifetimePredicate(
+                        predicate,
+                        lifetime,
+                        predicate.lifetimeParamBounds,
+                        lifetimeMap
+                    ) ?: continue@loop
+                    val name = lifetime.name ?: continue@loop
+                    addIfMissing(lifetimeToBounds, name)
+                    lifetimeToBounds[name]?.add(normalizePredicate(lifetimePredicate, name))
+                }
+            }
+        }
+
+        fun mapBounds(map: Map<String, Set<String>>, key: String): String? {
+            val bounds = map[key] ?: return null
+            if (bounds.isEmpty()) return null
+            return "$key: ${bounds.sorted().joinToString(" + ")}"
+        }
+
+        val bounds = lifetimeToBounds.keys.sorted().mapNotNull { mapBounds(lifetimeToBounds, it) } +
+            parameterToBounds.keys.sorted().mapNotNull { mapBounds(parameterToBounds, it) }
+
+        return if (bounds.isNotEmpty()) {
+            bounds.joinToString(separator = ",", prefix = " where ")
+        } else {
+            ""
+        }
+    }
+
+    companion object {
+        /**
+         * Recursively finds parent `RsGenericDeclarations` and collects all of their type parameters and
+         * their constraints.
+         */
+        fun create(context: PsiElement): GenericConstraints {
+            val typeParameters = mutableSetOf<RsTypeParameter>()
+            val lifetimes = mutableSetOf<RsLifetimeParameter>()
+            val constParameters = mutableSetOf<RsConstParameter>()
+            val whereClauses = mutableListOf<RsWhereClause>()
+
+            var genericDecl: RsGenericDeclaration? = (context as? RsGenericDeclaration) ?: context.parentOfType()
+            while (genericDecl != null) {
```

Comment:
In Rust, parent generics usually can't be used in nested items. For example, this is incorrect code:
```rust
fn foo<T>() {
    fn bar() -> T { unimplemented!() }
}
```
The 2 exceptions are trait end impl. So it's enough to look at `RsAbstrctable.owner`.
But I'm not sure it leads to any bugs. Maybe in a case like this:
```rust
fn foo<T>() where T: Foo {
    fn bar<T>() where T: Bar { unimplemented!() }
}
```

Reply:
Thanky, I added a check and a test for this.